osquery/config/parsers/decorators.cpp

```diff
@@ -0,0 +1,246 @@
+  // No need to define a key for the ::getData API.
+  return Status(0, "OK");
+}
+
+Status DecoratorsConfigParserPlugin::update(const std::string& source,
+                                            const ParserConfig& config) {
+  clearSources(source);
+  clearDecorations(source);
+  if (config.count(PARSER_NAME) > 0) {
+    // Each of these methods acquires the decorator lock separately.
+    // The run decorators method is designed to have call sites throughout
+    // the code base.
+    updateDecorations(source, config.at(PARSER_NAME));
+    runDecorators(DECORATE_LOAD, 0, source);
+  }
+
+  return Status(0, "OK");
+}
+
+void DecoratorsConfigParserPlugin::clearSources(const std::string& source) {
+  // Reset the internal data store.
+  WriteLock lock(DecoratorsConfigParserPlugin::kDecorationsMutex);
+  intervals_[source].clear();
+  always_[source].clear();
+  load_[source].clear();
+}
+
+void DecoratorsConfigParserPlugin::updateDecorations(
+    const std::string& source, const pt::ptree& decorators) {
+  WriteLock lock(DecoratorsConfigParserPlugin::kDecorationsMutex);
+  // Assign load decorators.
+  auto& load_key = kDecorationPointKeys.at(DECORATE_LOAD);
+  if (decorators.count(load_key) > 0) {
+    for (const auto& item : decorators.get_child(load_key)) {
+      load_[source].push_back(item.second.data());
+    }
+  }
+
+  // Assign always decorators.
+  auto& always_key = kDecorationPointKeys.at(DECORATE_ALWAYS);
+  if (decorators.count(always_key) > 0) {
+    for (const auto& item : decorators.get_child(always_key)) {
+      always_[source].push_back(item.second.data());
+    }
+  }
+
+  // Check if intervals are defined.
+  auto& interval_key = kDecorationPointKeys.at(DECORATE_INTERVAL);
+  if (decorators.count(interval_key) > 0) {
+    auto& interval = decorators.get_child(interval_key);
+    for (const auto& item : interval) {
+      size_t rate = std::stoll(item.first);
+      if (rate % 60 != 0) {
+        LOG(WARNING) << "Invalid decorator interval rate " << rate
+                     << " in config source: " << source;
+        continue;
+      }
+
+      // This is a valid interval, update the set of intervals to include
+      // this value. When intervals are checked this set is scanned, if a
+      // match is found, then the associated config data is executed.
+      for (const auto& interval_query : item.second) {
+        intervals_[source][rate].push_back(interval_query.second.data());
+      }
+    }
+  }
+}
+
+inline void addDecoration(const std::string& source,
+                          const std::string& name,
+                          const std::string& value) {
+  DecoratorsConfigParserPlugin::kDecorations[source][name] = value;
+}
+
+inline void runDecorators(const std::string& source,
+                          const std::vector<std::string>& queries) {
+  for (const auto& query : queries) {
+    auto results = SQL(query);
+    if (results.rows().size() > 0) {
```

Comment:
My thought is to provide results and warn on the undefined behavior, otherwise the warning should be promoted to an error. I am assuming there are cases where 1 row is returned during testing and at some point in the future multiple rows are returned-- I do not want this restriction to break flows but rather raise attention.